### ISSUE TITLE
[BUG] Probe: Reduce load on the host apiserver during namespace/secret update

### ISSUE THREAD

**Issue by amshuman-kr** [gardener/dependency-watchdog#15]
**Describe the bug:**
The dependency watchdog probe watches/reconciles the namespaces and secrets to re-configure/setup probes for the individual shoot namespaces to be probed.

Such reconciliation is not optimal right now as it treats each update to the namespace/secret as potentially a change to the kubeconfig secrets and proceeds to explicitly retrieve the secrets from the host apiserver and re-configure the probes.

This is suboptimal and can cause high load on the host apiserver and might end up throttling (see the log below) the probe process itself leading to many of the namespaces' shoot apiserver not probed for substantial period of time (> `2m`) which could be enough time for the individual `kube-controller-manager`s to act (potentially erroneously due to apiserver loadbalancer issues) on unresponsive nodes and mark them as `NotReady`.

**Expected behaviour:**
Individual probes for the different namespaces' shoot apiservers should be up as long as possible and should be reconfigured only if the kubeconfig secrets actually changed (with the content of the kubeconfig changing).

The down time for the individual probes should be as little as possible with secrets being loaded from the local informer cache rather than using a get call on the host apiserver (thus also reducing the load on the host apiserver).

**How To Reproduce (as minimally and precisely as possible):**

**Logs:**
```
Throttling request took 196.193411ms, request: GET:https://...:443/api/v1/namespaces/.../secrets/dependency-watchdog-external-probe
```
**Screenshots (if applicable):**

**Anything else we need to know?**:

**Comment by vlerenc** [comment]
Thanks for opening @danielfoehrKn and @amshuman-kr in regards to the misbehaviour noticed last week.

Another remark: Since we have very large secrets and you mentioned all data would be pulled, this could be responsible not only for significant load, but also for significant network traffic and be another major cost driver.

/critical
Raising priority as this problem has caused actual issues due to the KCM marking nodes as `NotReady` and the consequences of that.

**Comment by danielfoehrKn** [comment]
Maybe another question from my side: why is there an internal and also external probe using the LoadBalancer?  Just want to understand the use-case.
To my understanding that adds quite some external network traffic (have not measured it though).

@vlerenc isn't the network traffic only cluster-internal for pulling the secrets?

**Comment by amshuman-kr** [comment]
> Maybe another question from my side: why is there an internal and also external probe using the LoadBalancer? Just want to understand the use-case.
To my understanding that adds quite some external network traffic (have not measured it though).

The external probe probes the shoot apiserver via the loadbalancer (like the shoot node's kubelet does) and the internal probe probes the shoot apiserver via the service cluster IP. This is to be sure to scale down KCM only if the issue is due to the loadbalancer and not something in the apiserver pods themselves.

**Comment by amshuman-kr** [comment]
> @vlerenc isn't the network traffic only cluster-internal for pulling the secrets?

The secrets are pulled from the seed apiserver which sits in the soil cluster. Hence, it will be external network traffic.

**Comment by amshuman-kr** [comment]
> Raising priority as this problem has caused actual issues due to the KCM marking nodes as NotReady and the consequences of that.

Agreed in the priority. I have started work on this already.

**Comment by vlerenc** [comment]
Yes @amshuman-kr and we have seen that e.g. the gardener resource manager and similar reconcilers are responsible for some fraction of the exorbitant network traffic. That may be another one of interest that we may want to get down.

**Comment by amshuman-kr** [comment]
@vlerenc Just to be clear, the get calls happen only whenever the shoot namespaces and secrets in them are updated. So, it doesn't happen for each probe by the watchdog. So, traffic reduction may not be substantial. Or might be substantial only during seed updates. But this is a critical fix anyway to make sure the probes are minimally disrupted during seed updates.

**Comment by amshuman-kr** [comment]
Fixed in the release [v0.4.1](https://github.com/gardener/dependency-watchdog/releases/tag/v0.4.1).
